lib/phlex/component.rb

```diff
@@ -1,27 +1,112 @@
 # frozen_string_literal: true
+require "digest"
+require "byebug"
 
 module Phlex
   class Component
     include Node, Context
 
+    class << self
+      def cache_key
+        @cache_key ||= {
+          template: checksum,
+          subcomponents: subcomponents.map(&:checksum),
+          ancestors: dependant_ancestor_checksums
+        }
+      end
+
+      def checksum
+        @checksum ||= Digest::MD5.hexdigest(file)
+      end
+
+      def subcomponents
+        @subcomponents ||= direct_subcomponents.reduce(Set.new) do |set, component|
+          set << component
+          component.subcomponents.each { set << _1 }
+          set
+        end
+      end
+
+      private
+
+      def dependant_ancestor_checksums
+        dependant_ancestors.map do |ancestor|
+          Digest::MD5.hexdigest(
+            File.read(
+              Module.const_source_location(ancestor.name).first
+            )
+          )
```

Comment:
We could make a local checksum cache so we only ever need to build a checksum once per file path.